File: cisco-provided/dlog_utils/perl/dlog_ascii.pl
#!/usr/bin/perl 
# dlog_ascii.pl - dump IronPort delivery logs ascii format
#
# Copyright (C) 2002, IronPort Systems.  All rights reserved.
# $Revision: 1.9 $

use IO::String;
use IronPort_DeliveryLog;

use strict vars;

#global version
our $VERSION = 1;
our $backwards_compat = 0;
our $revision = q$Name: DLOG_PARSING_TOOLS_1_4 $;

package DLOG_ASCII;

sub parse_split_string
{
	# This code is for reference.  An old version of the delivery logs
 	# may have embedded an code object in the report files instead of 	
	# a string.  This parses the "split-string" class back into a string

	my $r = shift;
	my $key;
	my $retV = "";

	if (ref($r) eq "HASH") {
		my $hr = $r->{__dict__};
		foreach $key (keys(%$hr)) { 
			my $val = $hr->{$key};

			if (ref($val) eq "SCALAR") {
				# print "Key $key => $val\n";
			}
			elsif (ref($val) eq "ARRAY") {
			 	#print "Key $key => array of ", @$val, "\n";
				return join("", @$val);
			}
		}

	}

	return $r;
}

sub callback 
{
	my ($self, $rec) = @_;

	$self->cb_start($rec)    if $rec->{Type} == $IronPort_DeliveryLog::START;
	$self->cb_delivery($rec) if $rec->{Type} == $IronPort_DeliveryLog::DELIVERY;
	$self->cb_bounce($rec)   if $rec->{Type} == $IronPort_DeliveryLog::BOUNCE;
	$self->cb_end($rec)      if $rec->{Type} == $IronPort_DeliveryLog::END;
}

sub flush
{
    my ($self) = @_;
	my $fh = $self->{FileHandle};
	my $str = $self->{IO}->string_ref;
	$self->{IO} = new IO::String;
	print $fh $$str;
}

sub print
{
	my ($self, @args) = @_;
	my $fh = $self->{IO};
	print $fh @args;
}

sub printf
{
	my ($self, @args) = @_;
	my $fh = $self->{IO};
	printf $fh @args;
}

sub cb_start
{
	my ($self, $rec) = @_;
	my $VER = $rec->{Version};
	my $log_tm = IronPort_DeliveryLog::Time_to_String($rec->{Log_Time});
	
	$self->{Version} = $VER;

	$self->print("START $log_tm\n");
    if ($backwards_compat) {
	    $self->print(". Version " . "" . "\n");
    }
    else {
	    $self->print(". Version " . $VERSION . "\n");
        $self->print(". File Version $VER\n");
    }
	$self->print("\n");
	$self->flush();
}

sub cb_end
{
	my ($self, $rec) = @_;
    if ($rec->{Log_Time}[0] == 0 && $rec->{Log_time}[1] == 0) {
        return;
    }
	my $log_tm = IronPort_DeliveryLog::Time_to_String($rec->{Log_Time});

    

	if ($self->{Processing_Finished} != 1) {
		$self->{Processing_Finished} = 1;
		$self->print("END $log_tm\n");
        if (!$backwards_compat) {
            $self->print( "\n");
        }
		$self->flush();
	}
}

sub _errs_to_str
{
	my $arr = shift;

	if ($#$arr < 0) { return "[]"; }
	else            { return "[" . join (", ", @$arr) . "]"; }
}

my $PAT = '[\x0-\x1F\7F-]';
sub ascii_format
{
	my $str_ref = shift;
	my $val;
	while ($$str_ref =~ /($PAT)/) 
	{
		my $chr = $1;
		my $val = sprintf ("\\%03o", ord($chr));
		$$str_ref =~ s/$chr/$val/g;
	}
	return $str_ref;
}

sub ip_str
{
	my $ip = shift;
	my ($d, $c, $b, $a) = ($ip & 0xFF, 
						   ($ip >> 8) & 0xFF,
						   ($ip >> 16) & 0xFF,
						   ($ip >> 24) & 0xFF);

	return sprintf("%u.%u.%u.%u", $a, $b, $c, $d);
}

sub cb_delivery
{
	my ($self, $rec) = @_;
	
	my $log_tm  = IronPort_DeliveryLog::Time_to_String($rec->{Log_Time});
	my $delv_tm = IronPort_DeliveryLog::Time_to_String($rec->{Time});

	$self->print("DELIVERY ", $log_tm, "\n");
	$self->print(". Time   ", $delv_tm, "\n");
	$self->print(". Bytes  ", $rec->{Bytes}, "\n");
	$self->print(". Msg_Id ", $rec->{Msg_Id}, "\n");
	$self->print(". IP     ", ip_str($rec->{IP}), "\n");
	$self->print(". From   ", ${ascii_format(\$rec->{From})}, "\n");
	$self->print(". Domain ", $rec->{Domain}, "\n");

	#handle V4 added fields
	if ($self->{Version} >= 4) {
		$self->print(". Src_IP ", ip_str($rec->{SRC_IP}), "\n");
		$self->print(". Code   ", $rec->{Code}, "\n");
		$self->print(". Reply  ", $rec->{Reply}, "\n");
	}
	
	$self->print("\n");

	$self->rcpt_dump ($rec->{Recipient_Information});

    if (!$backwards_compat) {
        $self->print("\n");
    }

	$self->cust_dump ($rec->{Customer_Information});
	$self->print("\n");
	$self->flush();
}

sub cb_bounce
{
	my ($self, $rec) = @_;
	my $err_tm = IronPort_DeliveryLog::Time_to_String($rec->{Time});
	my $log_tm = IronPort_DeliveryLog::Time_to_String($rec->{Log_Time});

	$self->print("BOUNCE ", $log_tm, "\n");
	$self->print(". Time   ", $err_tm, "\n");
	$self->print(". Bytes  ", $rec->{Bytes}, "\n");
	$self->print(". Msg_Id ", $rec->{Msg_Id}, "\n");
	$self->print(". IP     ", ip_str($rec->{IP}), "\n");
	$self->print(". From   ", ${ascii_format(\$rec->{From})}, "\n");
	$self->print(". Reason ", $rec->{Reason}, "\n");
	$self->print(". Code   ", $rec->{Code}, "\n");
	#handle V4 added field
	if ($self->{Version} >= 4) {
		$self->print(". Src_IP ", ip_str($rec->{SRC_IP}), "\n");
	}
	$self->print("\n");

	$self->rcpt_dump ($rec->{Recipient_Information});

    if (!$backwards_compat) {
        $self->print("\n");
    }

	$self->cust_dump ($rec->{Customer_Information});
	$self->err_dump ($rec->{Errors});
	$self->print("\n");
	$self->flush();
}

sub arr_size 
{
	my $arr = shift;
	return $#$arr + 1;
}

sub rcpt_dump
{
	my $self = shift;
	my $arr = shift;
	my $sz = arr_size($arr);

	if ($sz == 0) { return; }

	$self->printf("Rcpt Data: (%d elem)\n", $sz);
	foreach my $elem (@$arr)
	{
		my ($rid, $email, $attempts) = (
			$elem->{Rcpt_Id},
			$elem->{Address},
			$elem->{Attempt});

		ascii_format(\$email);
		$attempts += 1;
		$self->print(" .. Rcpt_Id $rid\n");
		$self->print(" ..   Attempt_Number $attempts\n");
		$self->print(" ..   Email $email\n");
	}
}

sub err_dump
{
	my $self= shift;
	my $arr = shift;
	my $sz = arr_size($arr);

	if ($sz > 0) 
	{
		$self->printf("\n");
		$self->printf("Error Data: (%d elem)\n", $sz);

		foreach my $elem (@$arr) 
		{
			$self->print(" .. " . $elem . "\n");
		}
	}
}

sub cust_dump
{
	my $self = shift;
	my $arr = shift;
	my $sz = arr_size($arr);

	if ($sz == 0) { return; }

	$self->printf("Cust Data: (%d elem)\n", $sz);
	foreach my $elem (@$arr)
	{
		my ($name, $val) = (
			$elem->{Name},
			$elem->{Value});
                #$val2 = parse_split_string($val);
                my $val2 = $val;
		$self->print(" .. $name => $val2 \n");
	}
}

sub new 
{
	my $proto = shift;
	my $class = ref ($proto) || $proto;
	my $fh = shift;

    my $self = {};

	if (! defined $fh) { $fh = \*STDOUT; }
	$self->{FileHandle} = $fh;
	$self->{IO} = new IO::String;

	bless $self, $class;
	return $self;
}

1;
package main;

sub callback 
{
	my $record = shift;

	my $str = 'UNKNOWN';
	$str = 'ERROR ' if $record->{Type} == $IronPort_DeliveryLog::ERROR;
	$str = 'DELVRY' if $record->{Type} == $IronPort_DeliveryLog::DELIVERY;
	$str = 'BOUNCE' if $record->{Type} == $IronPort_DeliveryLog::BOUNCE;
	$str = 'START ' if $record->{Type} == $IronPort_DeliveryLog::START;
	$str = 'END   ' if $record->{Type} == $IronPort_DeliveryLog::END;

	print "Record type: $str at ";
	print IronPort_DeliveryLog::Time_to_String($record->{Log_Time});
	print "\n";
}

my $next = 0;
my @infiles = ();
*OUTFILE = \*STDOUT;

sub print_version
{
     $revision =~ /.*: (.*)/;
     print "\n\tdlog_ascii.pl version $1\n\n";

}


while (my $arg = shift (@ARGV)) {
   if ($next) {
       open OUTFILE, "> $arg" || die "Unable to open $arg for output\n";
       $next = 0;
   }
   elsif ($arg eq "-version") {
       print_version();
       exit(0);
   }
   elsif ($arg eq "-o") {
      $next = 1;
   }
   elsif ($arg eq "-B") {
      $backwards_compat = 1;
   }
   else {
      push @infiles, $arg;
   }
}

if ($#infiles < 0) {
   die "Usage:  $0  [ -o output_file ] [-B] dlog_1 dlog_2 .. dlog_n\n";
}

my $parser = new IronPort_DeliveryLog;
my $xml_printer = new DLOG_ASCII(\*OUTFILE);
my $cref = sub { $xml_printer->callback(@_); };


$parser->set_callback($cref);

if ($backwards_compat) {
    $parser->ignore('end');
}

foreach my $file (@infiles) 
{
	$parser->read_file($file);
    if ($backwards_compat) {
	    $parser->ignore('start');
    }

}

if ($backwards_compat) {
    my $rec = { Log_Time=>[time, 0] };
    $xml_printer->cb_end($rec);
}
close OUTFILE;

1;

File: cisco-provided/dlog_utils/c/dlog_ascii.c
/*
 * dlog_parser.c - dump IronPort delivery logs in CSV, XML and ascii
 *
 * Copyright (C) 2002, IronPort Systems.  All rights reserved.
 * $Revision: 1.2 $
 */

#ifdef DO_MEMCHECK
/* 
 * memcheck checks for bounds reads/writes, allocation errors, etc.
 *
 * export MEMCHECK_REUSE = 1
 * if processing large files, o.w. memory will just grow
 */
#include <sys/types.h>
#include <memcheck.h>
#endif

#include <stdarg.h> /* problematic for some OSes? */

#include <sys/types.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <time.h>
#include <string.h>
#include <stdlib.h>

#include "dlog_common.h"
#include "un_pickle.h"


#define MAX_SUPPORTED_VERSION 4

extern FILE * outfile; /* = stdout; */
char* revision = "$Name: DLOG_PARSING_TOOLS_1_4 $";

#if 0
extern start_rec_t start_record;
extern end_rec_t end_record;
extern delv_rec_t delv_record;
extern bounce_rec_t bounce_record;
#endif
extern int file_version;


void
dump_errors_normal(str_info_t *arr, int count)
{
   int idx;

   if (count == 0) { return; }

   fprintf (outfile, "\nError Data: (%d elem)\n", count);

   for (idx = 0; idx < count; idx++) {
	  fprintf (outfile, " .. %*.*s\n", arr[idx].len, arr[idx].len, arr[idx].val);
   }
}

void
dump_errors_bc(str_info_t *arr, int count)
{
   int idx;

   if (count == 0) { return; }

   fprintf (outfile, "\nError Data (%d elem)\n", count);

   for (idx = 0; idx < count; idx++) {
	  fprintf (outfile, " .. %*.*s\n", arr[idx].len, arr[idx].len, arr[idx].val);
   }
}

void(*dump_errors)(str_info_t *, int);

static char *
ascii_print(int len, char *src)
{
   char *_chr, *end, tmp;
   int pos = 0;
   char* buf = next_pbuf();

   end = (len > BSZ) ? src + (BSZ - 1) : src + len;

   for (_chr = src; _chr != end; _chr++)
   {
	  tmp = *_chr;
	  if (tmp < 40 || tmp >= 127 || tmp == '\\')
	  {
		 /* fprintf(outfile, "%d {%*.*s}\n", len, len, len, src); */
		 pos += sprintf(buf + pos, "\\%03o", tmp);
	  }
	  else {
		 buf[pos++] = tmp;
	  }
   }
   buf[pos] = '\0';

   return buf;
	  
}

void 
dump_rcpt(rcpt_info_t *arr, int count)
{
   int idx;

   if (count == 0) { return; }

   fprintf (outfile, "\nRcpt Data: (%d elem)\n", count);

   for (idx = 0; idx < count; idx++) {
	  fprintf (outfile, " .. Rcpt_Id %d\n", arr[idx].rcpt_id);
	  fprintf (outfile, " ..   Attempt_Number %d\n", arr[idx].attempt);
	  fprintf (outfile, " ..   Email %s\n", 
			  ascii_print(arr[idx].address.len, arr[idx].address.val));
   }
}


void 
dump_cust(cust_info_t *arr, int count)
{
   int idx;

   if (count == 0) { return; }

   fprintf (outfile, "\nCust Data: (%d elem)\n", count);

   for (idx = 0; idx < count; idx++) {
	  fprintf (outfile, " .. %*.*s => %*.*s\n", 
			  arr[idx].name.len,
			  arr[idx].name.len,
			  arr[idx].name.val,
			  arr[idx].value.len,
			  arr[idx].value.len,
			  arr[idx].value.val);
   }
}


int
dump_start_record_local(start_rec_t *rec)
{
   printf ("opening up\n");
   assert(rec && rec->magic == MAGIC_START);
   fprintf (outfile, "START %s\n", render_time(rec->log_secs, rec->log_usecs));
   fprintf (outfile, ". Version %d\n", rec->version);
   fprintf (outfile, ". File Version %d\n", rec->file_version);
   fprintf (outfile, "\n");
   return 0;
}

int
dump_delv_record_local(delv_rec_t *rec)
{
   assert(rec && rec->magic == MAGIC_DELV);
   fprintf (outfile, "DELIVERY %s\n", render_time(rec->log_secs, rec->log_usecs));
   fprintf (outfile, ". Time   %s\n", render_time(rec->entry_secs, rec->entry_usecs));
   fprintf (outfile, ". Bytes  %u\n", rec->bytes);
   fprintf (outfile, ". Msg_Id %u\n", rec->mesg_id);
   fprintf (outfile, ". IP     %s\n", render_ip(rec->ip_addr));
   fprintf (outfile, ". From   %*.*s\n", rec->from.len, rec->from.len, rec->from.val);
   fprintf (outfile, ". Domain %*.*s\n", rec->domain.len, rec->domain.len, rec->domain.val);

   /* handle v4 record additions */
   if (file_version >= 4) {
   	fprintf (outfile, ". Src_IP %s\n", render_ip(rec->src_ip));
   	fprintf (outfile, ". Code   %*.*s\n", rec->code.len, rec->code.len, rec->code.val);
   	fprintf (outfile, ". Reply  %*.*s\n", rec->reply.len, rec->reply.len, rec->reply.val);
   } 
   dump_rcpt(rec->rcpt_arr, rec->n_rcpt);
   dump_cust(rec->cust_arr, rec->n_cust);
   fprintf (outfile, "\n");
   return 0;
}

int
dump_bounce_record_local(bounce_rec_t *rec)
{
   assert(rec && rec->magic == MAGIC_BOUNCE);
   fprintf (outfile, "BOUNCE %s\n", render_time(rec->log_secs, rec->log_usecs));
   fprintf (outfile, ". Time   %s\n", render_time(rec->entry_secs, rec->entry_usecs));
   fprintf (outfile, ". Bytes  %u\n", rec->bytes);
   fprintf (outfile, ". Msg_Id %u\n", rec->mesg_id);
   fprintf (outfile, ". IP     %s\n", render_ip(rec->ip_addr));
   fprintf (outfile, ". From   %*.*s\n", rec->from.len, rec->from.len, rec->from.val);
   fprintf (outfile, ". Reason %*.*s\n", rec->reason.len, rec->reason.len, rec->reason.val);
   fprintf (outfile, ". Code   %*.*s\n", rec->code.len, rec->code.len, rec->code.val);

   /* handle v4 added fields */  
   if (file_version >= 4) {
   	fprintf (outfile, ". Src_IP %s\n", render_ip(rec->src_ip));
   }
   dump_rcpt(rec->rcpt_arr, rec->n_rcpt);
   dump_cust(rec->cust_arr, rec->n_cust);
   dump_errors(rec->err_arr, rec->n_err);
   fprintf (outfile, "\n");
   return 0;
}


int
dump_end_record_local(end_rec_t *rec)
{
   assert(rec && rec->magic == MAGIC_END);
   fprintf (outfile, "END %s\n", render_time(rec->log_secs, rec->log_usecs));
   fprintf (outfile, "\n");
   return 0;
}

int (*dump_delv_record)(delv_rec_t *rec);
int (*dump_bounce_record)(bounce_rec_t *rec);
int (*dump_start_record)(start_rec_t *rec);
int (*dump_end_record)(end_rec_t *rec);

void start_report(int opts[]) { 
   dump_errors = opts[BACKWARDS_COMPAT] ? &dump_errors_bc : &dump_errors_normal;
   dump_start_record = dump_start_record_local;
   dump_end_record = dump_end_record_local;
   dump_delv_record = dump_delv_record_local;
   dump_bounce_record = dump_bounce_record_local;
}
 
void end_report() { }


void usage(const char *name)
{
   fprintf (stderr, "Usage: %s [ options ] log0 log1 ...\n", name);
   fprintf (stderr, "     -o               output file\n");
   fprintf (stderr, "     -H               don't display hard bounce messages\n");
   fprintf (stderr, "     -D               don't display delivery messages\n");
   fprintf (stderr, "     -version         version information\n");
   fprintf (stderr, "     -B               backwards compatability\n");
   fprintf (stderr, "     -h               this message\n");
   exit(1);
}

